File: src/main/java/com/youngjun/auth/core/domain/token/TokenReader.java
package com.youngjun.auth.core.domain.token;

import com.youngjun.auth.api.support.error.AuthException;
import com.youngjun.auth.api.support.error.ErrorType;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class TokenReader {

    private final TokenRepository tokenRepository;

    private final TokenParser tokenParser;

    public TokenReader(TokenRepository tokenRepository, TokenParser tokenParser) {
        this.tokenRepository = tokenRepository;
        this.tokenParser = tokenParser;
    }

    public Token readVerified(RefreshToken refreshToken) {
        tokenParser.verify(refreshToken);
        List<Token> tokens = tokenRepository.read(refreshToken);
        if (tokens.isEmpty()) {
            throw new AuthException(ErrorType.TOKEN_NOT_FOUND_ERROR, null);
        }
        return tokens.get(0);
    }

}

File: src/main/kotlin/com/youngjun/auth/core/domain/token/TokenReader.kt
package com.youngjun.auth.core.domain.token

import com.youngjun.auth.core.api.support.error.AuthException
import com.youngjun.auth.core.api.support.error.ErrorType.TOKEN_NOT_FOUND_ERROR
import com.youngjun.auth.storage.db.core.token.TokenRepository
import org.springframework.stereotype.Component

@Component
class TokenReader(
    private val tokenRepository: TokenRepository,
) {
    fun read(refreshToken: RefreshToken): Token {
        val tokens = tokenRepository.read(refreshToken)
        if (tokens.isEmpty()) {
            throw AuthException(TOKEN_NOT_FOUND_ERROR)
        }
        return tokens.first()
    }
}
